napari/_qt/threading.py

```diff
@@ -0,0 +1,663 @@
+import inspect
+from functools import wraps
+from typing import Type, Dict, Callable, Any, Set, Optional
+from qtpy.QtCore import QObject, QThread, Signal, Slot, QRunnable, QThreadPool
+import time
+import re
+
+
+def as_generator_function(func: Callable) -> Callable:
+    """Turns a regular function (single return) into a generator function."""
+
+    @wraps(func)
+    def genwrapper(*args, **kwargs):
+        yield
+        return func(*args, **kwargs)
+
+    return genwrapper
+
+
+class WorkerBase(QRunnable, QObject):
+    """Base class for creating a Worker that can run in another thread."""
+
+    started = Signal()  # emitted when the work is started
+    finished = Signal()  # emitted when the work is finished
+    returned = Signal(object)  # emitted with return value
+    errored = Signal(object)  # emitted with error object on Exception
+
+    def __init__(self, *args, **kwargs) -> None:
+        QRunnable.__init__(self)
+        QObject.__init__(self)
+        self._abort_requested = False
+        self._running = False
+
+    def quit(self) -> None:
+        """Send a message to abort the worker."""
+        self._abort_requested = True
+
+    @property
+    def abort_requested(self) -> bool:
+        """Whether the worker has been requested to stop."""
+        return self._abort_requested
+
+    @property
+    def is_running(self) -> bool:
+        """Whether the worker has been started"""
+        return self._running
+
+    @Slot()
+    def run(self):
+        """Start the worker.
+
+        The end-user should never need to call this function.
+        But it cannot be made private or renamed, since it is called by Qt.
+
+        **This** is the function that actually gets called when calling
+        :func:`QThreadPool.start(worker)`.  It simply wraps the :meth:`work`
+        method, and emits a few signals.  Subclasses should NOT override this
+        method (except with good reason), and instead should implement
+        :meth:`work`.
+        """
+        self.started.emit()
+        self._running = True
+        try:
+            self.work()
+        except Exception as exc:
+            self.errored.emit(exc)
+        self.finished.emit()
+
+    def work(self):
```

Comment:
btw in which order slot and base implementation will be called once work implementation from child class is called @tlambert03 ?

Reply:
yeah, that should be in the docstring!

1. `QThreadPool.globalInstance().start(worker)` calls `Worker.run()`
2. `Worker.run` calls `Worker.work` (implemented by subclasses).

(as a convenience... `Worker.start()` is there to calls that first QThreadPool method)

Reply:
gotcha. thanks.

Reply:
add docstring to both `worker.run` and `worker.start`